mod/state-transition/pkg/core/state_processor_staking.go

```diff
@@ -171,6 +173,128 @@ func (sp *StateProcessor[
 		math.Gwei(sp.cs.MaxEffectiveBalance()),
 	)
 
+	// BeaconKit enforces a cap on the validator set size. If the deposit
+	// breaches the cap, we find the validator with the smallest stake and
+	// mark it as withdrawable so that it will be evicted next epoch and
+	// its deposits returned.
+
+	if sp.processingGenesis {
+		// minor optimization: we do check in Genesis that
+		// validators do no exceed cap. So hitting cap here
+		// should never happen
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	nextEpochVals, err := sp.nextEpochValidatorSet(st)
+	if err != nil {
+		return err
+	}
+	//#nosec:G701 // no overflow risk here
+	if uint32(len(nextEpochVals)) < sp.cs.GetValidatorSetCap() {
+		// cap not hit, just add the validator
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// Adding the validator would breach the cap. Find the validator
+	// with the smallest stake among current and candidate validators
+	// and kick it out.
+	smallestVal, err := sp.smallest(nextEpochVals)
+	if err != nil {
+		return err
+	}
+
+	slot, err := st.GetSlot()
+	if err != nil {
+		return err
+	}
+	nextEpoch := sp.cs.SlotToEpoch(slot) + 1
+
+	if val.GetEffectiveBalance() <= smallestVal.GetEffectiveBalance() {
+		// in case of tie-break among candidate validator we prefer
+		// existing one so we mark candidate as withdrawable
+		// We wait next epoch to return funds, as a way to curb spamming
+		val.SetWithdrawableEpoch(nextEpoch)
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// mark exiting validator for eviction and add candidate
+	smallestVal.SetWithdrawableEpoch(nextEpoch)
+	idx, err := st.ValidatorIndexByPubkey(smallestVal.GetPubkey())
+	if err != nil {
+		return err
+	}
+	if err = st.UpdateValidatorAtIndex(idx, smallestVal); err != nil {
+		return err
+	}
+	return sp.addValidatorInternal(st, val, dep.GetAmount())
+}
+
+// nextEpochValidatorSet returns the current estimation of what next epoch
+// validator set would be.
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, _, _, _, _, _, _, _, ValidatorT, _, _, _, _,
+]) nextEpochValidatorSet(st BeaconStateT) ([]ValidatorT, error) {
+	slot, err := st.GetSlot()
+	if err != nil {
+		return nil, err
+	}
+	withdrawEpoch := sp.cs.SlotToEpoch(slot) + 1
+
+	vals, err := st.GetValidators()
+	if err != nil {
+		return nil, err
+	}
+	activeVals := make([]ValidatorT, 0, len(vals))
+	for _, val := range vals {
+		if val.GetEffectiveBalance() <= math.U64(sp.cs.EjectionBalance()) {
+			continue
+		}
+		if val.GetWithdrawableEpoch() == withdrawEpoch {
+			continue
+		}
+		activeVals = append(activeVals, val)
+	}
+
+	return activeVals, nil
+}
+
+// TODO: consider moving this to BeaconState directly
+func (sp *StateProcessor[
+	_, _, _, _, _, _, _, _, _, _, _, _, ValidatorT, _, _, _, _,
+]) smallest(currentVals []ValidatorT) (
```

Comment:
rename to `lowestStakeVal()` or smth similar

Reply:
Tricky to get rid of the receiver (which I agree is not needed) and get the typing right. We have other methods here that could be free functions and are not for this reason.
I would rather defer everything to a general cleanup of the typing system